Nodejs/Product/Nodejs/Intellisense/ExpansionClient.cs

```diff
@@ -0,0 +1,308 @@
+                        string defaultValue;
+                        if (ErrorHandler.Succeeded(_session.GetFieldValue(decl, out defaultValue))) {
+                            templateText = templateText.Replace("$" + decl + "$", defaultValue);
+                        }
+                    }
+                    templateText = templateText.Replace("$end$", "");
+
+                    // we can finally figure out where the selected text began witin the original template...
+                    int selectedIndex = templateText.IndexOf("$selected$");
+                    if (selectedIndex != -1) {
+                        var selection = _textView.Selection;
+
+                        // now we need to get the indentation of the $selected$ element within the template,
+                        // as we'll need to indent the selected code to that level.
+                        string indentation = GetTemplateSelectionIndentation(templateText, selectedIndex);
+
+                        var start = _selectionStart.GetPosition(_textView.TextBuffer.CurrentSnapshot);
+                        var end = _selectionEnd.GetPosition(_textView.TextBuffer.CurrentSnapshot);
+                        if (end < start) {
+                            // we didn't actually have a selction, and our negative tracking pushed us
+                            // back to the start of the buffer...
+                            end = start;
+                        }
+                        var selectedSpan = Span.FromBounds(start, end);
+
+                          if (surroundsWithStatement &&
+                            String.IsNullOrWhiteSpace(_textView.TextBuffer.CurrentSnapshot.GetText(selectedSpan))) {
+
+                            // Surround With can be invoked with no selection, but on a line with some text.
+                            // In that case we need to inject an extra new line.
+                            var endLine = _textView.TextBuffer.CurrentSnapshot.GetLineFromPosition(end);
+                            var endText = endLine.GetText().Substring(end - endLine.Start);
+                            if (!String.IsNullOrWhiteSpace(endText)) {
+                                edit.Insert(end, _textView.Options.GetNewLineCharacter());
+                            }
+                        }
+
+                        IndentSpan(
+                            edit,
+                            indentation,
+                            _textView.TextBuffer.CurrentSnapshot.GetLineFromPosition(start).LineNumber + 1, // 1st line is already indented
+                            _textView.TextBuffer.CurrentSnapshot.GetLineFromPosition(end).LineNumber
+                        );
+                    }
+                }
+
+                // we now need to update any code which was not selected  that we just inserted.
+                IndentSpan(edit, baseIndentation, ts[0].iStartLine + 1, ts[0].iEndLine);
+
+                edit.Apply();

```

Comment:
The snippets won't behave very well when different formatting options are supplied. E.g. tabs vs. spaces, or 2 spaces rather than four spaces. If we're lucky, VS might handle things for us if we supply tabs in the snippets rather than spaces, but otherwise we should add some additional handling logic here.